donut/modules/voting/helpers.py

```diff
@@ -0,0 +1,395 @@
+
+    group = form['group']
+    group = int(group) if group else None
+
+    params = {
+        'title': form['title'].strip(),
+        'description': form['description'].strip() or None,
+        'start_time': start,
+        'end_time': end,
+        'auth': 'auth' in form,
+        'public': 'public' in form,
+        'group_id': group
+    }
+    if survey_id:
+        update_survey_params(survey_id, params)
+    else:
+        access_key = generate_random_string(ACCESS_KEY_LENGTH)
+        params['access_key'] = access_key
+        params['creator'] = get_user_id(flask.session['username'])
+        query = 'INSERT INTO surveys (' + ', '.join(
+            params) + ') VALUES (' + ', '.join(['%s'] * len(params)) + ')'
+        with flask.g.pymysql_db.cursor() as cursor:
+            cursor.execute(query, list(params.values()))
+    return flask.redirect(
+        flask.url_for('voting.edit_questions', access_key=access_key))
+
+
+def get_survey_params(survey_id):
+    query = """
+        SELECT title, description, start_time, end_time, group_id, auth, public
+        FROM surveys WHERE survey_id = %s
+    """
+    with flask.g.pymysql_db.cursor() as cursor:
+        cursor.execute(query, [survey_id])
+        return cursor.fetchone()
+
+
+def update_survey_params(survey_id, params):
+    query = 'UPDATE surveys SET ' + ', '.join(
+        key + ' = %s' for key in params) + ' WHERE survey_id = %s'
+    with flask.g.pymysql_db.cursor() as cursor:
+        cursor.execute(query, [*params.values(), survey_id])
+
+
+def set_questions(survey_id, questions):
+    delete_query = 'DELETE FROM survey_questions WHERE survey_id = %s'
+    insert_question_query = """
+        INSERT INTO survey_questions (survey_id, title, description, list_order, type_id)
+        VALUES (%s, %s, %s, %s, %s)
+    """
+    insert_choice_query = """
+        INSERT INTO survey_question_choices (question_id, choice)
+        VALUES (%s, %s)
+    """
+    with flask.g.pymysql_db.cursor() as cursor:
+        cursor.execute(delete_query, [survey_id])
+        for order, question in enumerate(questions):
+            cursor.execute(insert_question_query, [
+                survey_id, question['title'], question['description'].strip()
+                or None, order, question['type']
+            ])
+            question_id = cursor.lastrowid
+            for choice in question.get('choices', []):
+                cursor.execute(insert_choice_query, [question_id, choice])
+
+
+def get_my_surveys(user_id):
+    query = """
+        SELECT title, description, access_key,
+        start_time, start_time >= NOW() AS unopened,
+        end_time <= NOW() as closed, end_time
+        FROM surveys WHERE creator = %s
+    """
+    with flask.g.pymysql_db.cursor() as cursor:
+        cursor.execute(query, [user_id])
+        return cursor.fetchall()
+
+
+def delete_survey(survey_id):
+    query = 'DELETE FROM surveys WHERE survey_id = %s'
```

Comment:
if possible please also delete associated `survey_responses` and `survey_question_choices` since there's literally no reason they should be in the database if the survey has been deleted.

Reply:
jk you did it below with cascades